aten/src/ATen/native/sparse/SparseMatMul.cpp

```diff
@@ -0,0 +1,255 @@
+
+    The matrices `A` and 'B' should be in proper CSR structure, and their dimensions
+    should be compatible.
+
+    Inputs:
+      `n_row`         - number of row in A
+      `n_col`         - number of columns in B
+      `Ap[n_row+1]`   - row pointer
+      `Aj[nnz(A)]`    - column indices
+      `Ax[nnz(A)]     - nonzeros
+      `Bp[?]`         - row pointer
+      `Bj[nnz(B)]`    - column indices
+      `Bx[nnz(B)]`    - nonzeros
+    Outputs:
+      `Cp[n_row+1]` - row pointer
+      `Cj[nnz(C)]`  - column indices
+      `Cx[nnz(C)]`  - nonzeros
+
+    Note:
+      Output arrays Cp, Cj, and Cx must be preallocated
+  */
+  std::vector<int64_t> next(n_col, -1);
+  std::vector<scalar_t> sums(n_col, 0);
+
+  int64_t nnz = 0;
+
+  Cp[0] = 0;
+
+  for (int64_t i = 0; i < n_row; i++) {
+    int64_t head = -2;
+    int64_t length = 0;
+
+    int64_t jj_start = Ap[i];
+    int64_t jj_end = Ap[i + 1];
+    for (int64_t jj = jj_start; jj < jj_end; jj++) {
+      int64_t j = Aj[jj];
+      scalar_t v = Ax[jj];
+
+      int64_t kk_start = Bp[j];
+      int64_t kk_end = Bp[j + 1];
+      for (int64_t kk = kk_start; kk < kk_end; kk++) {
+        int64_t k = Bj[kk];
+
+        sums[k] += v * Bx[kk];
+
+        if (next[k] == -1) {
+          next[k] = head;
+          head = k;
+          length++;
+        }
+      }
+    }
+
+    for (int64_t jj = 0; jj < length; jj++) {
+      Cj[nnz] = head;
+      Cx[nnz] = sums[head];
+      nnz++;
+
+      int64_t temp = head;
+      head = next[head];
+
+      next[temp] = -1; // clear arrays
+      sums[temp] = 0;
+    }
+
+    Cp[i + 1] = nnz;
+  }
+}
+
+
+template <typename scalar_t>
+void sparse_matmul_kernel(
+    Tensor& output,
+    const Tensor& mat1,
+    const Tensor& mat2) {
+  /* 
+    Computes  the sparse-sparse matrix multiplication between `mat1` and `mat2`, which are sparse tensors in COO format. 
+  */
+
+  auto M = mat1.size(0);
+  auto K = mat1.size(1);
+  auto N = mat2.size(1);
+
+  auto mat1_indices_ = mat1._indices().contiguous();
+  auto mat1_values = mat1._values().contiguous();
+  LongTensor mat1_row_indices = mat1_indices_.select(0, 0);
+  LongTensor mat1_col_indices = mat1_indices_.select(0, 1);
+
+  Tensor mat1_indptr = coo_to_csr(mat1_row_indices.data_ptr<int64_t>(), M, mat1._nnz());
+
+  auto mat2_indices_ = mat2._indices().contiguous();
+  auto mat2_values = mat2._values().contiguous();
+  LongTensor mat2_row_indices = mat2_indices_.select(0, 0);
+  LongTensor mat2_col_indices = mat2_indices_.select(0, 1);
+
+  Tensor mat2_indptr = coo_to_csr(mat2_row_indices.data_ptr<int64_t>(), K, mat2._nnz()); 
+
+  auto nnz = _csr_matmult_maxnnz(M, N, mat1_indptr.data_ptr<int64_t>(), mat1_col_indices.data_ptr<int64_t>(), 
+      mat2_indptr.data_ptr<int64_t>(), mat2_col_indices.data_ptr<int64_t>());
+
+  auto output_indices = output._indices();
+  auto output_values = output._values();
+
+  Tensor output_indptr = at::empty({M + 1}, kLong);
+
+  output_indices.resize_({2, nnz});
```

Comment:
Can we use resize_output to throw a warning if the given output is the incorrect size (unless it's empty)?

Reply:
Sure